### DISCUSSION TITLE
API Monetisation

### DISCUSSION THREAD

**Discussion by GurdipS5** [dotnet/yarp#1478] [Q&A] [answered]
### What should we add or change to make your life better?

Hi, I am looking into API monetisation for my APIs. Thi would mean charging (i.e. per invocation, or per month, usage tiers, etc).

### Why is this important to you?

This would allow me to use reverse proxy for all my req's. If thi out of the scope of a reverse proxy, what would you recommend?

**Comment by Tratcher** [comment, marked as answer]
YARP doesn't provide this functionality by default, but you could use YARP to add this functionality to an existing application.

Monitization requires two main parts: A) Authentication/Authorization, and B) Reporting.

A) You need to have an authentication and authorization mechanism on every request so you know who to charge, and to refuse requests if their account isn't in good standing. See https://microsoft.github.io/reverse-proxy/articles/authn-authz.html

B) You need to report each request (preferably in batches for perf reasons) to the billing system. This could be done in custom middleware https://microsoft.github.io/reverse-proxy/articles/middleware.html.

You don't need YARP to implement A or B, you could choose to implement them directly in the application. However, you could choose to use YARP to add these capabilities in front of an existing app if you wanted to keep the logic separate.
